### PR TITLE
ISEAR Emotions Processor

### PR DESCRIPTION
**Pull request by tomaryancit**
@lapolonio thanks for the reply. 
I'll just try to explain where I am so far:
I have created a fine-tuned model with the isear.csv dataset following your instructions and that all works good.
I am now trying to serve the model and use the run_app.py to return predictions.
The saved_model and variables I have downloaded locally and when I run a do_predict locally on that model the results look accurate. I will post some examples shortly, but basically the sentences in test.tsv would be like 'I am very happy' and 'I am very sad' and the 7 float values returned would match under the label types. That all works good with do_predict.

However when i use the model in serving from the run_app.py (hardcoding a test example like 'I am very sad'. I seem to get the same prediction, regardless of what I send in. The values are all pretty even, but the Joy label comes out tops EVERY time.

If you could have a look at the run_classifier.py and point me in the right direction with what I am doing wrong I would really appreciate it.
Thank you,
Tom

### CODE REVIEW

**Comment by lapolonio** [comment]
looks alright to me.

it probably has to do with how you are actually calling the model. How are you running predictions?

Also check out https://github.com/tomaryancit/bert/blob/a5b8ead400e91a3b3dbb67295e17583d714869c4/run_app.py#L60 for how to get the pretty label.

**Comment by tomaryancit** [comment]
@lapolonio I am running predictions with this script:

--------------------------------------------------------------------------------------------------------
#!/bin/bash     
TRAINED_CLASSIFIER="C:\models\best_model_ckpt\model.ckpt-581"
ISEAR_DIR="C:\models\isear_data"
BERT_BASE_DIR="C:\models\bert_base\uncased_L-12_H-768_A-12"
OUTPUT_DIR="C:\models\outputs"
OUTPUT_FILE="C:\models\outputs\test_results.tsv"

**python run_classifier.py --task_name=isear --do_predict=true --data_dir=$ISEAR_DIR --vocab_file=$BERT_BASE_DIR/vocab.txt --bert_config_file=$BERT_BASE_DIR/bert_config.json --init_checkpoint=$TRAINED_CLASSIFIER --max_seq_length=128 --output_dir=$OUTPUT_DIR**
--------------------------------------------------------------------------------------------------------
**test,tsv** is:
index	sentence
0	i am so angry
0	i am extremely sad
0	i am extremely happy
0	i am very unhappy

**test_results.tsv** I get are:
0.8718204	0.017259441	0.010415434	0.06656713	0.0014592747	0.015705448	0.016772918
0.0068754684	0.0051930887	0.005554631	0.018795557	0.004932479	0.9541948	0.0044539524
0.0024607785	0.004653761	0.0027854743	0.0022160264	0.9713467	0.007915857	0.008621418
0.018595515	0.009295957	0.00865367	0.10277452	0.007821805	0.8311814	0.021677122

Emotions are:
anger	disgust	fear	guilt	joy	sadness	shame

These test_results are accurate,
**All good so far!**

Then I run the following with **do_serve=True** which gives me a saved_model.pb:
**ft_estimator = TensorFlow(source_directory=project_folder,
                    vm_size='STANDARD_NC6',
                    script_params = {
                          '--task_name': 'isear',
                          '--do_train' : True,
                          '--do_serve': True,
                          '--data_dir': ds.path('isear_data_path/').as_mount(),
                          '--vocab_file': ds.path('BERT_model/uncased_L-12_H-768_A-12/vocab.txt').as_mount(),
                          '--bert_config_file': ds.path('BERT_model/uncased_L-12_H-768_A-12/bert_config.json').as_mount(),
                          '--init_checkpoint': ds.path('BERT_model/uncased_L-12_H-768_A-12/bert_model.ckpt').as_mount(),
                          '--max_seq_length': 128,
                          '--train_batch_size': 32,
                          '--learning_rate': 0.00015152,
                          '--num_train_epochs': 3.0,
                          '--output_dir': './outputs'
                    },
                    entry_script='run_classifier.py',
                    use_gpu=True)**

I put the saved_model.pb into C:\models\model_6\ and when I run 
**saved_model_cli show --dir C:\models\model_6\1554648672 --all**
I get

MetaGraphDef with tag-set: 'serve' contains the following SignatureDefs:

signature_def['serving_default']:
  The given SavedModel SignatureDef contains the following input(s):
    inputs['examples'] tensor_info:
        dtype: DT_STRING
        shape: (-1)
        name: input_example_tensor:0
  The given SavedModel SignatureDef contains the following output(s):
    outputs['probabilities'] tensor_info:
        dtype: DT_FLOAT
        shape: (-1, 7)
        name: loss/Softmax:0
  Method name is: tensorflow/serving/predict
PS C:\models>

I then run tensorflow/serving using the docker image ( https://www.tensorflow.org/tfx/serving/docker):

**docker run -p 8500:8500 -p 8501:8501 --mount type=bind,source=c:\models\model_6,target=/models/model_6 -e MODEL_NAME=model_6 -t tensorflow/serving**

Then I use a version of your run_app.py to run locally on port 5000.
I then send requests in POSTMAN but **I get the same label every time!**

Output of run_app.py is here:

   Use a production WSGI server instead.
 * Debug mode: on
 * Restarting with stat
 * Debugger is active!
 * Debugger PIN: 197-615-899
 * Running on http://127.0.0.1:5000/ (Press CTRL+C to quit)
['anger', 'disgust', 'fear', 'guilt', 'joy', 'sadness', 'shame']
I am sad
[2019-04-12 19:03:41,381] INFO in run_app: Predicted Label: joy
127.0.0.1 - - [12/Apr/2019 19:03:41] "?[37mPOST / HTTP/1.1?[0m" 200 -
['anger', 'disgust', 'fear', 'guilt', 'joy', 'sadness', 'shame']
I am angry
[2019-04-12 19:03:56,029] INFO in run_app: Predicted Label: joy
127.0.0.1 - - [12/Apr/2019 19:03:56] "?[37mPOST / HTTP/1.1?[0m" 200 -


I'm stuck at this stage and would really appreciate some guidance!
Thanks.


run_app.py is:
--------------------------------------------------------------------------------------------

import grpc
import tensorflow as tf

import run_classifier as classifiers
import tokenization

from tensorflow_serving.apis import predict_pb2
from tensorflow_serving.apis import prediction_service_pb2_grpc
from tensorflow.core.framework import tensor_pb2
from tensorflow.core.framework import tensor_shape_pb2
from tensorflow.core.framework import types_pb2

from flask import Flask
from flask import request

import random

app = Flask(__name__)


@app.route("/", methods = ['GET'])
def hello():
  return "Hello BERT predicting your emotions! Try posting a string to this url"


@app.route("/", methods = ['POST'])
def predict():
  # MODEL PARAMS
  max_seq_length = 128

  channel = grpc.insecure_channel("localhost:8500")   #("bert-agnews:8500")
  stub = prediction_service_pb2_grpc.PredictionServiceStub(channel)

  # Parse Description
  tokenizer = tokenization.FullTokenizer(
    vocab_file="asset/vocab.txt", do_lower_case=True)
  processor = classifiers.IsearProcessor()
  label_list = processor.get_labels()
  print(label_list)
  content = request.get_json()
  request_id = str(random.randint(1, 9223372036854775807))
  inputExample = processor._create_example([request_id, content['description']], 'test')
  tf_example = classifiers.from_record_to_tf_example(3, inputExample, label_list, max_seq_length, tokenizer)
  model_input = tf_example.SerializeToString()

  # Send request
  # See prediction_service.proto for gRPC request/response details.
  model_request = predict_pb2.PredictRequest()
  model_request.model_spec.name = 'model_6'
  model_request.model_spec.signature_name = 'serving_default'
  dims = [tensor_shape_pb2.TensorShapeProto.Dim(size=1)]
  tensor_shape_proto = tensor_shape_pb2.TensorShapeProto(dim=dims)
  tensor_proto = tensor_pb2.TensorProto(
    dtype=types_pb2.DT_STRING,
    tensor_shape=tensor_shape_proto,
    string_val=[model_input])

#  print([model_input])
  print(content['description'])
  model_request.inputs['examples'].CopyFrom(tensor_proto)
  result = stub.Predict(model_request, 10.0)  # 10 secs timeout
  result = tf.make_ndarray(result.outputs["probabilities"])
  pretty_result = "Predicted Label: " + label_list[result[0].argmax(axis=0)]
  app.logger.info("Predicted Label: %s", label_list[result[0].argmax(axis=0)])
  return pretty_result


if __name__ == '__main__':
  app.run(debug=True, host='127.0.0.1')

**Comment by tomaryancit** [comment]
Don't mean to be annoying you @lapolonio but would you get a chance to have a quick look at the method I am using above for calling the model and doing predictions? 

Just that I am stuck on this and would really appreciate some advice or help to move forward.
Thank you.

**Comment by lapolonio** [comment]
I think when you export the init_checkpoint should be `"C:\models\best_model_ckpt\model.ckpt-581"` currently you have `BERT_model/uncased_L-12_H-768_A-12/bert_model.ckpt`

**Comment by tomaryancit** [comment]
Thanks @lapolonio I made that change to use model.ckpt-581 as the init_checkpoint.
This created a new saved_model.pb which unfortunately is giving the same problem as before - i.e. prediction is always 'joy', regardless of utterance.

Output from run_app.py :

127.0.0.1 - - [20/Apr/2019 11:11:25] "?[37mPOST / HTTP/1.1?[0m" 200 -
['anger', 'disgust', 'fear', 'guilt', 'joy', 'sadness', 'shame']
utterance: **i am so angry**
result: **[[0.13873199 0.14457144 0.13604145 0.1460678  0.14955397 0.138183
  0.14685035]]**
[2019-04-20 11:13:25,623] INFO in run_app: Predicted Label: **joy**
127.0.0.1 - - [20/Apr/2019 11:13:25] "?[37mPOST / HTTP/1.1?[0m" 200 -
['anger', 'disgust', 'fear', 'guilt', 'joy', 'sadness', 'shame']
utterance: **i am extremely happy**
result: **[[0.13873194 0.14457141 0.13604145 0.1460678  0.14955392 0.13818304
  0.14685036]]**
[2019-04-20 11:13:50,502] INFO in run_app: Predicted Label: **joy**
127.0.0.1 - - [20/Apr/2019 11:13:50] "?[37mPOST / HTTP/1.1?[0m" 200 -

**Comment by lapolonio** [comment]
Are the predictions the same or different than before?

You have to isolate whether the issue is the model or the scoring
code. Are you using the original serving function?

**Comment by tomaryancit** [comment]
@lapolonio the predictions are the same as before I made the init_checkpoint change.
They are always similar to the results I provided above.
The scores can differ ever so slightly, as you can see from the two separate result scores listed, but they are always very close to these scores.
You can see that the third last score in each case is always slightly the highest, giving a classification of 'joy'.
I have not changed the serving function that i posted up originally in run_classifier.py

Just a thought - would it be anything to do with the fact that we supply the set_type as 'test' in the _create_example() method or the fact that we do not supply 'is_real_example' in the features? 
I'm grasping at straws a bit here!

**Comment by lapolonio** [comment]
is_real_examples is handled here:
https://github.com/lapolonio/bert/blob/master/run_classifier.py#L639

So your serving code is the following?
https://github.com/lapolonio/bert/blob/master/run_classifier.py#L980

the label is generated so no need to specify.

**Comment by tomaryancit** [comment]
@lapolonio serving code has some slight differences, but i think it does the same?


 if FLAGS.do_serve:

    feature_spec = {
        "input_ids": tf.FixedLenFeature([FLAGS.max_seq_length], tf.int64),
        "input_mask": tf.FixedLenFeature([FLAGS.max_seq_length], tf.int64),
        "segment_ids": tf.FixedLenFeature([FLAGS.max_seq_length], tf.int64),
        "label_ids": tf.FixedLenFeature([], tf.int64),
        "is_real_example": tf.FixedLenFeature([], tf.int64)
      }

    def serving_input_receiver_fn():      
      serialized_tf_example = tf.placeholder(dtype=tf.string,
                                              shape=[None],
                                              name='input_example_tensor')
      receiver_tensors = {'examples': serialized_tf_example}
      features = tf.parse_example(serialized_tf_example, feature_spec)

      return tf.estimator.export.ServingInputReceiver(features, receiver_tensors)


    tf.logging.info("***** Running new serve *****")

    estimator._export_to_tpu = False  # this is important
    path = estimator.export_savedmodel(FLAGS.output_dir, serving_input_receiver_fn)
    print(path)

if __name__ == "__main__":
  flags.mark_flag_as_required("data_dir")
  flags.mark_flag_as_required("task_name")
  flags.mark_flag_as_required("vocab_file")
  flags.mark_flag_as_required("bert_config_file")
  flags.mark_flag_as_required("output_dir")
  tf.app.run()

**Comment by lapolonio** [comment]
it's best to change as little as possible while you are working out bugs.

**Comment by lapolonio** [comment]
This has an example of executing the model without the tensorflow serving container. http://shzhangji.com/blog/2018/05/14/serve-tensorflow-estimator-with-savedmodel/

**Comment by tomaryancit** [comment]
thanks for this @lapolonio , i will try this example.
